crates/pyndakaas/src/lib.rs

```diff
@@ -1,107 +1,713 @@
+					.map(|any| any.and_then(|lit| lit.extract::<Lit>()))
+					.collect::<PyResult<_>>()?;
+				let mut guard = self.0.lock().unwrap();
+				guard
+					.add_clause(clause.into_iter().map(|lit| lit.0))
+					.unwrap();
+				Ok(())
+			}
+
+			fn add_encoding(&mut self, con: ConstraintArg, enc: Option<Encoder>) -> PyResult<()> {
+				let mut guard = self.0.lock().unwrap();
+				encode_constraint(&mut *guard, con, enc)
+			}
+
+			#[new]
+			fn new() -> Self {
+				Self(Default::default())
+			}
+
+			fn new_vars(&mut self, num_vars: usize) -> Vec<Lit> {
+				let mut guard = self.0.lock().unwrap();
+				guard
+					.new_var_range(num_vars)
+					.into_iter()
+					.map(|lit| Lit(lit.into()))
+					.collect()
+			}
+
+			fn set_time_limit(&mut self, limit: Option<Duration>) {
+				let mut guard = self.0.lock().unwrap();
+				guard.set_terminate_callback(limit.map(dur_term_fn))
+			}
+
+			fn solve_assuming(&self, assumptions: Vec<Lit>) -> (Status, HashMap<i32, bool>) {
+				let mut guard = self.0.lock().unwrap();
+				let vars = guard.emitted_vars();
+				match guard.solve_assuming(assumptions.iter().map(|&lit| lit.0)) {
+					SolveResult::Satisfied(sol) => (
+						Status::SATISFIED,
+						vars.into_iter()
+							.map(|var| (var.into(), sol.value(var.into())))
+							.collect(),
+					),
+					SolveResult::Unsatisfiable(fail) => (
+						Status::UNSATISFIABLE,
+						assumptions
+							.iter()
+							.map(|&lit| (lit.0.into(), fail.fail(lit.0)))
+							.collect(),
+					),
+					SolveResult::Unknown => (Status::UNKNOWN, HashMap::new()),
+				}
+			}
+		}
+
+		#[pymethods]
+		impl KissatInner {
+			fn add_clause(&mut self, clause: Bound<'_, PyIterator>) -> PyResult<()> {
+				let clause: Vec<Lit> = clause
+					.into_iter()
+					.map(|any| any.and_then(|lit| lit.extract::<Lit>()))
+					.collect::<PyResult<_>>()?;
+				let mut guard = self.0.lock().unwrap();
+				guard
+					.add_clause(clause.into_iter().map(|lit| lit.0))
+					.unwrap();
+				Ok(())
+			}
+
+			fn add_encoding(&mut self, con: ConstraintArg, enc: Option<Encoder>) -> PyResult<()> {
+				let mut guard = self.0.lock().unwrap();
+				encode_constraint(&mut *guard, con, enc)
+			}
+
+			#[new]
+			fn new() -> Self {
+				Self(Default::default())
+			}
+
+			fn new_vars(&mut self, num_vars: usize) -> Vec<Lit> {
+				let mut guard = self.0.lock().unwrap();
+				guard
+					.new_var_range(num_vars)
+					.into_iter()
+					.map(|lit| Lit(lit.into()))
+					.collect()
+			}
+
+			fn set_time_limit(&mut self, limit: Option<Duration>) {
+				let mut guard = self.0.lock().unwrap();
+				guard.set_terminate_callback(limit.map(dur_term_fn))
+			}
+
+			fn solve_assuming(
+				&self,
+				assumptions: Vec<Lit>,
+			) -> PyResult<(Status, HashMap<i32, bool>)> {
+				if !assumptions.is_empty() {
+					return Err(PyNotImplementedError::new_err(
+						"Kissat does not support assumptions",
+					));
+				}
+				let mut guard = self.0.lock().unwrap();
+				let vars = guard.emitted_vars();
```

Comment:
might this also expose the auxiliary variables of encodings?

Reply:
Yes, we currently don't distinguish between them. We could track the Python emitted ones I suppose and then create the variables from that list.